plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/expression/ExpEmbedding.kt

```diff
@@ -360,13 +360,26 @@ data class FieldAccess(val receiver: ExpEmbedding, val field: FieldEmbedding) :
 
     override fun toViperStoringIn(result: VariableEmbedding, ctx: LinearizationContext) {
         val receiverViper = receiver.toViper(ctx)
-        val fieldAccess = PrimitiveFieldAccess(ExpWrapper(receiverViper, receiver.type), field)
-        val invariant = accessInvariant?.fillHole(ExpWrapper(receiverViper, receiver.type))?.pureToViper(toBuiltin = true, ctx.source)
+        val receiverWrapper = ExpWrapper(receiverViper, receiver.type)
+        // If the field is immutable, it is necessary to unfold predicates
+        if (field.unfoldToAccess) unfoldHierarchy(receiverWrapper, ctx)
+
+        val fieldAccess = PrimitiveFieldAccess(receiverWrapper, field)
+        val invariant = accessInvariant?.fillHole(receiverWrapper)?.pureToViper(toBuiltin = true, ctx.source)
         invariant?.let { ctx.addStatement(Stmt.Inhale(it, ctx.source.asPosition)) }
         ctx.addStatement(Stmt.assign(result.toViper(ctx), fieldAccess.pureToViper(toBuiltin = false, ctx.source), ctx.source.asPosition))
         invariant?.let { ctx.addStatement(Stmt.Exhale(it, ctx.source.asPosition)) }
     }
 
+    private fun unfoldHierarchy(receiverWrapper: ExpEmbedding, ctx: LinearizationContext) {
+        val hierarchyPath = (receiver.type as? ClassTypeEmbedding)?.hierarchyUnfoldPath(field.name)
+        hierarchyPath?.forEach {
+            val predAcc = it.predicateAccessInvariant().fillHole(receiverWrapper)
+                .pureToViper(toBuiltin = true, ctx.source) as? Exp.PredicateAccess
+            predAcc?.let { x -> ctx.addStatement(Stmt.Unfold(x)) }
+        }
+    }
```

Comment:
nit: the `x` in the last lambda is easy to guess the meaning of, much harder than what is being iterated over with the `forEach`.  Could you name the parameter of the `forEach` lambda instead of naming `x`?